### PR TITLE
Log messages refactors

### PR DESCRIPTION
**Pull request by leio10**
This PR changes the election statuses and mutations names to adapt the code to the actual election messages definition.

It also makes the `start_vote` and `end_vote` (previously known as `open_ballot_box` and `close_ballot_box`) asynchronous to keep consistency with the other actions (`start_key_ceremony` and `start_tally`). 

Finally, it implements a working protocol (not secure!) to the Dummy voting scheme to be able to test better the whole process.

### CODE REVIEW

**Review by beagleknight — APPROVED** [review]
Oh wow, what a great job @leio10 ! Calling this a refactor is not fair haha. Anyway I didn't see anything worth to change but I am just worried about some magic strings in some places that could be constants. We can do that in a later refactor, no worries!

File: decidim-bulletin_board-app/app/services/voting_scheme/dummy.rb [hunk 1]
```diff
@@ -1,9 +1,19 @@
...
+  # similar to a real voting scheme implementation. It works like this:
+  # - every Trustee choose a random number as their public election key: Ki = rand(50, 249) * 2 + 1
+  #  - these keys should be odd numbers between 100 and 500
+  # - the joint key is the product of those keys: JK = K1 * K2 * ... * Kn
+  # - voters encrypt each answer multiplying the joint election key by a random number and summing 0 or 1 to that: Aj = JK * rand(100, 500) + (0 | 1)
+  # - when the tally begins, the BB creates the tally cast summing all the encrypted votes per answer: At = A1 + A2 + ... + Am
+  # - each trustee calculates the modulus between each answer cast and their election public key and multiply the result by their election public key:
+  #   Si = (At % Ki) * Ki
+  # - the final decryption of each answer consists on multiplying all the shares for that answer, dividing it by the joint election key and then
+  #   calculating the inverse power with the number of trustees: R = ((S1 * S2 * ... * Sn) / JK) ** (1/n)
```
Comment (agustibr) [inline decidim-bulletin_board-app/app/services/voting_scheme/dummy.rb:15]: 😍

**Review by agustibr — APPROVED** [review]
👍 great!

File: decidim-bulletin_board-app/app/services/voting_scheme/dummy.rb [hunk 2]
```diff
@@ -1,9 +1,19 @@
 # frozen_string_literal: true
 
-require "prime"
-
 module VotingScheme
-  # A dummy implementation of a voting scheme, only for tests purposes
+  # A dummy implementation of a voting scheme, only for tests purposes.
+  # It uses a very basic math to perform simple but unsecure encryption operations,
+  # similar to a real voting scheme implementation. It works like this:
+  # - every Trustee choose a random number as their public election key: Ki = rand(50, 249) * 2 + 1
```
Comment (slickepinne) [inline decidim-bulletin_board-app/app/services/voting_scheme/dummy.rb:7]:
```suggestion
  # - every Trustee chooses a random number as their public election key: Ki = rand(50, 249) * 2 + 1
```

File: decidim-bulletin_board-app/spec/commands/start_key_ceremony_spec.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+  it "creates the log entry for the message" do
+    expect { subject }.to change(LogEntry, :count).by(1)
+  end
+
+  it "doesn't change the election voting scheme state" do
+    expect { subject }.not_to(change(election, :voting_scheme_state))
+  end
+
+  it "change the election status to key_ceremony" do
```
Comment (slickepinne) [inline decidim-bulletin_board-app/spec/commands/start_key_ceremony_spec.rb:39]:
```suggestion
  it "changes the election status to key_ceremony" do
```

File: decidim-bulletin_board-app/spec/commands/start_key_ceremony_spec.rb [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    end
+  end
+
+  context "when the client is not the election authority" do
+    let(:client) { create(:authority, private_key: private_key) }
+    let(:private_key) { generate(:private_key) }
+
+    it_behaves_like "starting the key ceremony fails"
+
+    it "broadcast invalid" do
```
Comment (slickepinne) [inline decidim-bulletin_board-app/spec/commands/start_key_ceremony_spec.rb:61]:
```suggestion
    it "broadcasts invalid" do
```

File: decidim-bulletin_board-app/spec/commands/start_key_ceremony_spec.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    end
+  end
+
+  context "when the client is a trustee" do
+    let(:client) { Trustee.first }
+    let(:private_key) { Test::PrivateKeys.trustees_private_keys.first }
+
+    it_behaves_like "starting the key ceremony fails"
+
+    it "broadcast invalid" do
```
Comment (slickepinne) [inline decidim-bulletin_board-app/spec/commands/start_key_ceremony_spec.rb:72]:
```suggestion
    it "broadcasts invalid" do
```

File: decidim-bulletin_board-app/spec/commands/start_key_ceremony_spec.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      expect { subject }.to broadcast(:invalid, "Invalid client")
+    end
+  end
+
+  context "when the message author is not the authority" do
+    let(:extra_message_params) { { authority: create(:authority) } }
+
+    it_behaves_like "starting the key ceremony fails"
+
+    it "broadcast invalid" do
```
Comment (slickepinne) [inline decidim-bulletin_board-app/spec/commands/start_key_ceremony_spec.rb:82]:
```suggestion
    it "broadcasts invalid" do
```

File: decidim-bulletin_board-app/spec/jobs/start_key_ceremony_job_spec.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+  it "processes the message" do
+    expect { subject }.to change { PendingMessage.last.status } .from("enqueued").to("accepted")
+  end
+
+  context "when the message was already processed" do
+    before { described_class.perform_now(pending_message.id) }
+
+    it "doesn't change the message status" do
+      expect { subject }.not_to change { PendingMessage.last.status } .from("accepted")
```
Comment (slickepinne) [inline decidim-bulletin_board-app/spec/jobs/start_key_ceremony_job_spec.rb:20]: Do we need the `.from("accepted")` part here as we are only interested that the status didn't get change at all?

File: decidim-bulletin_board-ruby/lib/decidim/bulletin_board/client.rb [hunk 8]
```diff
@@ -94,6 +99,13 @@ def publish_results(election_id)
         publish_results.call
       end
 
+      def get_status(election_id)
```
Comment (slickepinne) [inline decidim-bulletin_board-ruby/lib/decidim/bulletin_board/client.rb:102]: This is the old name of the `get_election_status` method and shouldn't be included again. 😊

File: decidim-bulletin_board-ruby/spec/decidim/bulletin_board/client_spec.rb [hunk 9]
```diff
@@ -38,52 +38,54 @@ module BulletinBoard
...
         context "when everything went ok" do
           before do
-            stub_wisper_publisher("Decidim::BulletinBoard::Authority::OpenBallotBox", :call, :ok, double(status: "vote"))
+            stub_wisper_publisher("Decidim::BulletinBoard::Authority::StartKeyCeremony", :call, :ok, double(status: "enqueued"))
           end
 
-          it "calls the OpenBallotBox command and returns the new status" do
-            election = subject.open_ballot_box(election_id)
-            expect(election.status).to eq("vote")
+          it "calls the StartKeyCeremony command and return the result" do
```
Comment (slickepinne) [inline decidim-bulletin_board-ruby/spec/decidim/bulletin_board/client_spec.rb:51]:
```suggestion
          it "calls the StartKeyCeremony command and returns the result" do
```

File: decidim-bulletin_board-ruby/spec/decidim/bulletin_board/client_spec.rb [hunk 10]
```diff
@@ -124,7 +127,33 @@ module BulletinBoard
...
+        subject { instance.end_vote(election_id) }
+
+        let(:election_id) { "test.1" }
+
+        context "when everything went ok" do
+          before do
+            stub_wisper_publisher("Decidim::BulletinBoard::Authority::EndVote", :call, :ok, double(status: "enqueued"))
+          end
+
+          it "calls the EndVote command and return the result" do
```
Comment (slickepinne) [inline decidim-bulletin_board-ruby/spec/decidim/bulletin_board/client_spec.rb:145]:
```suggestion
          it "calls the EndVote command and returns the result" do
```

File: decidim-bulletin_board-ruby/spec/decidim/bulletin_board/client_spec.rb [hunk 11]
```diff
@@ -174,22 +204,23 @@ module BulletinBoard
...
         context "when everything went ok" do
           before do
-            stub_wisper_publisher("Decidim::BulletinBoard::Authority::StartTally", :call, :ok, double(status: "tally"))
+            stub_wisper_publisher("Decidim::BulletinBoard::Authority::StartTally", :call, :ok, double(status: "enqueued"))
           end
 
-          it "calls the StartTally command and returns the new status" do
-            election = subject.start_tally(election_id)
-            expect(election.status).to eq("tally")
+          it "calls the StartTally command and return the result" do
```
Comment (slickepinne) [inline decidim-bulletin_board-ruby/spec/decidim/bulletin_board/client_spec.rb:222]:
```suggestion
          it "calls the StartTally command and returns the result" do
```

**Review by slickepinne — APPROVED** [review]
That is huge. Fantastic job! 🎉 Next to some very, very tiny notes, just be sure to remove the `get_election` method from the `client`. 😊

File: decidim-bulletin_board-ruby/lib/decidim/bulletin_board/client.rb (hunk 8 above)
Comment (leio10) [reply decidim-bulletin_board-ruby/lib/decidim/bulletin_board/client.rb:102]: It was a rebase issue :(

**Pull request merged**